### ISSUE TITLE
Diffgram License Agreement v1 (DLA) Fails to Follow Open Source Spirit

### ISSUE THREAD

**Issue by martinvahi** [diffgram/diffgram#270]
The [version 1](https://github.com/diffgram/diffgram/blob/34dde448dd4f2f0f5a9f5dbcacae0aef3c026c06/LICENSE.md) of the Diffgram License Agreement v1 (DLA) that is in use in 2021_09_02 forbids to redistribute derivative works to 3. parties and it even forbids to pick ideas from the Diffgram project and create a competing project with those ideas. Citations:

-  You will not, in any form, provide the Software or any subset of the Software to third parties. You will not host the Software for third parties to use, or in any way attempt to distribute or make available any substantial set of the features or functionality of the Software.
- You will not compete with the Licensor in the Training Data Software market.

Basically those 2 points make the software as restricted as closed source freeware with the exception that in stead of delivering the software in binary form the software is delivered in source form. Even some closed source freeware licenses allow to redistribute the software in an unmodified form and do not demand limits on competition.

I do think that those kind of license terms is a BUG, because nowadays many people are sufficiently experienced to notice those problems and the Diffgram project does aim to be an OPEN SOURCE project, not just some closed source project with the binaries having a slightly different delivery form, source form.

Thank You for reading this bug report.

**Comment by anthony-sarkis** [comment]
@martinvahi    Thanks for the note!

One example of the need for restriction is to prevent AWS et. el. from copying this and starting their own service.

We are open to improving it of course - do you have any ideas on ways we can make it more open, while still protecting ourselves so we can continue to grow and improve the software?

Is this more of a philosophical question?
Or is there some use case that you have in mind, that you feel is reasonable, that this is blocking?

Best
Anthony

**Comment by martinvahi** [comment]
> One example of the need for restriction is to prevent AWS et. el. from copying this and starting their own service.
>
> We are open to improving it of course - do you have any ideas on ways we can make it more open, while still protecting ourselves so we can continue to grow and improve the software?

Easier said than done of course, but the key phrase is ["business model canvas"](https://creately.com/blog/diagrams/business-model-canvas-explained/) _([archival_copy](https://archive.vn/f2yFR))_ and one of its fields in some versions of it is titled "Unfair Advantage". The old-fashioned way to block competitors was to use lawyers to just repress the competitors. A bit further back in history and in stead of lawyers the competitors might have been blocked by a group of horsemen with swords or in WW2 era mafia world, a group of men with machine guns. Works for sure, but is that the kind of world that we want? Shouldn't COMPETITION be based on outperforming other competitors in stead of inhibiting other competitors? For example, in the Olympic Games runners need to OUTRUN their competitors in stead of finding ways how to inhibit them end even the use of special drugs is considered bad practice, despite the fact that the drugs, secret blood transfers, etc. do enhance performance.


> Is this more of a philosophical question?

Both, philosophical and practical. After all, people do choose their profession and employer not just according to what salary is offered, but ACCORDING TO THEIR PERSONAL VALUES, which is totally philosophical matter. Except that at some point the philosophy has its roots in very practical issues like the idea that one should not kill one's neighbour just because one does not like one's neighbour eventually reduces to game theory and the fact that for actions there will be counter-actions at some point or just plain change of world state. I'm not a Christian, but I like to think this way that even the God and Russian dictators have a RESPONSIBILITY to NOT RUIN THINGS FOR THEMSELVES. Or what about child labor? A philosophical matter, but in modern world marketing the company that uses child labor will probably find it to be a very practical matter, when it comes to bottom line, sales of the products that were produced with child labor.

So, in my case, as of 2021_09_03 I like a kind of competition that is based on OUTPERFORMING COMPETITORS without setting artificial barriers to clients and without inhibiting competitors by technical means or by violence. The use of patents and copyright is based on supermafia/state based violence combined with setting restrictions on clients. 

So, the way to "protect" Yourself from Amazon is to OUTPERFORM Amazon and Amazon is a really big and clumsy megacorporation that CAN NOT AFFORD to offer certain kinds of services that a smaller company can offer, INDEPENDENT OF THE AMOUNT OF MONEY THAT THEY HAVE FOR THEIR INVESTMENTS. It's basically like chess or 3D-first-person-shooter games, where one tries to choose a weapon/button that differs from the opponent's weapons/buttons, but that has a greater advantage in the given environment than the opponent's weapons/buttons have. Which strategy to choose is a bit like asking, how to play chess and someone like Kasparov is certainly much better at it than I am, but my observation is that the Kasparov used a set of ready-made solutions, had his own notes from various chess study books (or so he told at one interview that I'm currently too lazy to look up for reference) and if someone changes the rules of the chess buttons, then probably the notes and chess study books would have become at least party obsolete and the Kasparov would have been in trouble. With business models there seems to be a same kind of pattern: people use old heuristics for assembling new business models. For example, inthe 90-ties people thought that the closed source licensing model works, therefore a "proper software company" has to sell closed source software licenses like the hyper-rich Microsoft did, except that at some point RedHat and IBM started to earn a lot of money with services and as of 2021 even the Microsoft tries to upgrade its revenue portfolio by increasing service related percentage of the revenue. The Bing advertisement revenue is also a form of service revenue. 

OK, so first step, in the 3D-first-person-shooter game style, let's see, how Your "weapons" and the "weapons" of Amazon differ:

- Amazon has a lot of money, Diffgram probably not that much, so no competing in areas, where bigger investment allows to reduce the "unit cost" of the service.
- Amazon is a slowly reacting megacorporation, Diffgram is probably currently still small enough to be properly flexible, so adapting new technology is something that Diffgram can probably do faster and better AT A CONTINUOUS RATE than Amazon.
- Amazon CAN NOT AFFORD to study small niche markets, but Diffgram might do that to see common patterns in different very small niche markets, which in turn may lead to a discovery of a market that is totally new and that is at first too small for a megacorporation like Amazon to find profitable, but that delivers percentage wise a very nice margin for a smaller company. What exactly, remains to be studies, but I have difficulties imagining that Amazon would be able to offer consulting services with specialized software (data science related software or integration services) to anybody, specially hospitals, science teams, etc. For example, hospitals have professional mathematicians working on patient data to find treatment inefficiencies and possible flaws and medicine side effects and those mathematicians WILL NOT accept the kind of quality that Amazon is able to deliver.
- With the IoT trend there is the need to normalize data. I'm not sure that Amazon is the company that will ever be able to offer that kind of data conversion software development services to all those IoT device manufacturers. They can buy a startup that has a batch of conversion software modules, but after a little while they need to buy another startup that does the same thing and after that yet another startup, etc.  Amazon is not the company that flexibly delivers custom solutions. 
- etc. You can complete the list Yourself, You have Your own brains.

The general scheme is that startups look for new business models and once they find some business model that works, then some megacorporation just acquires the startup, id est the startup founders get their pay during an "exit", which can be the acquisition by a megacorporation or sale of shares (IPO). A megacorporation then scales the business of the acquired startup company up, but a megacorporation will hardly ever create anything new themselves. They just acquire smaller companies. So, from that point of view, Amazon is not even Your main competitor, as long as You stay away from business models, where a lot of money for investment allows to considerably drive down the "unit cost" of the service or product.

Basically, the question, what to do, reduces to the question: **how does one want to live one's life?**
That in turn has the philosophy component in it. It's one's life philosophy that one has to take to account, when choosing one's lifestyle or the alternative is a lot of antidepressants, alcohol for some, even suicide for some. That's why I prefer to get by as a freelancer, despite the fact that I would probably earn a lot more at some corporate managerial position or at some managerial-techno-gray-zone like the role of a corporate "software architect". Or look at the "poor souls" that work at universities. Certainly they have the competence to hold a better paying jot at the industry, and yet, they choose salary wise the shoddier option.

That is to say, **a business model of a company must match the life philosophy of the founders of the company.** I'm not so well informed of Your life philosophy, so I can't recommend anything. I just laid out some of the patterns here that I know of, how to construct a business model. Again, You may joke that I'm just trolling here in stead of giving You useful feedback, but what can You say for countering, what I just said in this post?

> Or is there some use case that you have in mind, that you feel is reasonable, that this is blocking?

As a freelancer I can not afford to spend time studying software that I can not use at my work due to nontechnical hurdles like licenses. A concrete use case for me is data analysis pipeline, where other people's work form some section of the pipe, but the DLA v1 forbids re-delivery of the software, which means that I can not use the Diffgram in software that I deliver to my clients and my clients in return can not use anything that contains Diffgram without having extra costs in addition to my service costs. Selling me a redistribution license or even giving it for free would not work either, because I want my clients to be free to use my creations as sub-components of whatever they are doing for whomever. 

You may also think of the situation this way that once upon a time, when the copyright law was established, a copyrighted work was quite self-containes. A book, a pice of music, a painting. Nowadays, software contains other software that contains other software that that contains other software that contains other software repeat_N_times("that contains other software"). The same with patents: a physical product depends on technology that depends on technology repeat_K_times("that depends on technology"), leaving to a mess, where the only parties benefiting from the patent system are those, who do not produce anything themselves (read: do not need any patent licenses themselves) or the ones with enough lawyers to squash competitors that have less lawyers and enough patents to fend off other well lawyered-up competitors that actually produce something and therefore potentially violate some patents. 

Even the Elon Musk has given a free license to all of the Tesla patents to all of the Tesla competitors just to increase the market of electric cars. A counter-example is Intel and AMD that squash competitors and live off of the fact that a lot of legacy software is highly dependent on the x86 and AMD64 CPUs. Think of C++, which looks like a CPU-independent language, but actually can contain [inline assembly](https://www.cs.uaf.edu/courses/cs301/2014-fall/notes/inline-assembly/), which might be used in game engines, some extra optimised math routines, etc. and those components can be sub-components of other software by being a dependency of a dependency of a dependency of a .... A new CPU vendor can not come to a market without supporting legacy software, but the only way to do that is to run the x86/AMD64 binaries, which need a license from Intel and AMD, which in turn can ask such a high license fee that the new CPU will not be price wise competitive on the market and as a result, the new CPUs get used only where old desktop software does not need to run like the Android phones, Raspberry_Pi_like_computers and alike. That is to say, patents can help business very much as they have helped Intel and AMD, but there's the question that does the business model of Intel and AMD fit one's life philosophy. As of 2021_09_03 certainly not with mine.

So, @anthony-sarkis , please write here, what Your life philosophy is and then I might try to assemble some new business model candidates here. Thank You for Your answer and thank You for reading my letter.

:-)

**Comment by anthony-sarkis** [comment]
@martinvahi  Thank you for all of the insights. let's continue this discussion over email. I can be reached at {firstname}@diffgram.com
